src/java.base/share/classes/sun/security/util/DerInputStream.java

```diff
@@ -57,518 +60,153 @@
     }
 
-    /**
-     * Get an enumerated from the input stream.
-     *
-     * @return the integer held in this DER input stream.
-     */
     public int getEnumerated() throws IOException {
-        if (buffer.read() != DerValue.tag_Enumerated) {
-            throw new IOException("DER input, Enumerated tag error");
-        }
-        return buffer.getInteger(getDefiniteLength(buffer));
+        return getDerValue().getEnumerated();
     }
 
-    /**
-     * Get a bit string from the input stream. Padded bits (if any)
-     * will be stripped off before the bit string is returned.
-     */
     public byte[] getBitString() throws IOException {
-        if (buffer.read() != DerValue.tag_BitString)
-            throw new IOException("DER input not an bit string");
-
-        return buffer.getBitString(getDefiniteLength(buffer));
+        return getDerValue().getBitString();
     }
 
-    /**
-     * Get a bit string from the input stream.  The bit string need
-     * not be byte-aligned.
-     */
     public BitArray getUnalignedBitString() throws IOException {
-        if (buffer.read() != DerValue.tag_BitString) {
-            throw new IOException("DER input not a bit string");
-        }
-
-        int length = getDefiniteLength(buffer);
-
-        if (length == 0) {
-            return new BitArray(0);
-        }
-
-        /*
-         * First byte = number of excess bits in the last octet of the
-         * representation.
-         */
-        length--;
-        int excessBits = buffer.read();
-        if (excessBits < 0) {
-            throw new IOException("Unused bits of bit string invalid");
-        }
-        int validBits = length*8 - excessBits;
-        if (validBits < 0) {
-            throw new IOException("Valid bits of bit string invalid");
-        }
-
-        byte[] repn = new byte[length];
-
-        if ((length != 0) && (buffer.read(repn) != length)) {
-            throw new IOException("Short read of DER bit string");
-        }
-
-        return new BitArray(validBits, repn);
+        return getDerValue().getUnalignedBitString();
     }
 
-    /**
-     * Returns an ASN.1 OCTET STRING from the input stream.
-     */
     public byte[] getOctetString() throws IOException {
-        if (buffer.read() != DerValue.tag_OctetString)
-            throw new IOException("DER input not an octet string");
-
-        int length = getDefiniteLength(buffer);
-        byte[] retval = new byte[length];
-        if ((length != 0) && (buffer.read(retval) != length))
-            throw new IOException("Short read of DER octet string");
-
-        return retval;
-    }
-
-    /**
-     * Returns the asked number of bytes from the input stream.
-     */
-    public void getBytes(byte[] val) throws IOException {
-        if ((val.length != 0) && (buffer.read(val) != val.length)) {
-            throw new IOException("Short read of DER octet string");
-        }
+        return getDerValue().getOctetString();
```

Comment:
For strict compatibility, should fail if getDerValue().isConstructed() is false.

There is no need to check this in other getXyz() calls because:
- For all primitive values (including various strings), we never allowed indefinite length and we never accepted constructed tags. Now we reject indefinite length if tag is not constructed.
- For SEQUENCE and SET, we always allow indefinite length when allowBER == true